// ngx-select-ex/esm5/ngx-select/ngx-select.interfaces.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/** @typedef {?} */
var TNgxSelectOptionType;
export { TNgxSelectOptionType };
/**
 * @record
 */
export function INgxSelectOptionBase() { }
/** @type {?} */
INgxSelectOptionBase.prototype.type;
/**
 * @record
 */
export function INgxSelectOption() { }
/** @type {?} */
INgxSelectOption.prototype.value;
/** @type {?} */
INgxSelectOption.prototype.text;
/** @type {?} */
INgxSelectOption.prototype.disabled;
/** @type {?} */
INgxSelectOption.prototype.data;
/**
 * @record
 */
export function INgxSelectOptGroup() { }
/** @type {?} */
INgxSelectOptGroup.prototype.label;
/** @type {?} */
INgxSelectOptGroup.prototype.options;
/**
 * @record
 */
export function INgxOptionNavigated() { }
/** @type {?} */
INgxOptionNavigated.prototype.index;
/** @type {?} */
INgxOptionNavigated.prototype.activeOption;
/** @type {?} */
INgxOptionNavigated.prototype.filteredOptionList;
/**
 * @record
 */
export function INgxSelectOptions() { }
/** @type {?|undefined} */
INgxSelectOptions.prototype.optionValueField;
/** @type {?|undefined} */
INgxSelectOptions.prototype.optionTextField;
/** @type {?|undefined} */
INgxSelectOptions.prototype.optGroupLabelField;
/** @type {?|undefined} */
INgxSelectOptions.prototype.optGroupOptionsField;
/** @type {?|undefined} */
INgxSelectOptions.prototype.multiple;
/** @type {?|undefined} */
INgxSelectOptions.prototype.allowClear;
/** @type {?|undefined} */
INgxSelectOptions.prototype.placeholder;
/** @type {?|undefined} */
INgxSelectOptions.prototype.noAutoComplete;
/** @type {?|undefined} */
INgxSelectOptions.prototype.disabled;
/** @type {?|undefined} */
INgxSelectOptions.prototype.autoSelectSingleOption;
/** @type {?|undefined} */
INgxSelectOptions.prototype.autoClearSearch;
/** @type {?|undefined} */
INgxSelectOptions.prototype.noResultsFound;
/** @type {?} */
INgxSelectOptions.prototype.keepSelectedItems;
/** @type {?|undefined} */
INgxSelectOptions.prototype.size;
/** @type {?|undefined} */
INgxSelectOptions.prototype.keyCodeToRemoveSelected;
/** @type {?|undefined} */
INgxSelectOptions.prototype.keyCodeToOptionsOpen;
/** @type {?|undefined} */
INgxSelectOptions.prototype.keyCodeToOptionsClose;
/** @type {?|undefined} */
INgxSelectOptions.prototype.keyCodeToOptionsSelect;
/** @type {?|undefined} */
INgxSelectOptions.prototype.keyCodeToNavigateFirst;
/** @type {?|undefined} */
INgxSelectOptions.prototype.keyCodeToNavigatePrevious;
/** @type {?|undefined} */
INgxSelectOptions.prototype.keyCodeToNavigateNext;
/** @type {?|undefined} */
INgxSelectOptions.prototype.keyCodeToNavigateLast;
/** @type {?|undefined} */
INgxSelectOptions.prototype.isFocused;
