// ngx-select-ex/esm5/ngx-select/ngx-select.classes.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import * as escapeStringNs from 'escape-string-regexp';
/** @type {?} */
var escapeString = escapeStringNs;
var NgxSelectOption = /** @class */ (function () {
    function NgxSelectOption(value, text, disabled, data, _parent) {
        if (_parent === void 0) { _parent = null; }
        this.value = value;
        this.text = text;
        this.disabled = disabled;
        this.data = data;
        this._parent = _parent;
        this.type = 'option';
        this.cacheRenderedText = null;
    }
    Object.defineProperty(NgxSelectOption.prototype, "parent", {
        get: /**
         * @return {?}
         */
        function () {
            return this._parent;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} sanitizer
     * @param {?} highlightText
     * @return {?}
     */
    NgxSelectOption.prototype.renderText = /**
     * @param {?} sanitizer
     * @param {?} highlightText
     * @return {?}
     */
    function (sanitizer, highlightText) {
        if (this.cacheHighlightText !== highlightText || this.cacheRenderedText === null) {
            this.cacheHighlightText = highlightText;
            if (this.cacheHighlightText) {
                this.cacheRenderedText = sanitizer.bypassSecurityTrustHtml((this.text + '').replace(new RegExp(escapeString(this.cacheHighlightText), 'gi'), '<strong>$&</strong>'));
            }
            else {
                this.cacheRenderedText = sanitizer.bypassSecurityTrustHtml(this.text);
            }
        }
        return this.cacheRenderedText;
    };
    return NgxSelectOption;
}());
export { NgxSelectOption };
if (false) {
    /** @type {?} */
    NgxSelectOption.prototype.type;
    /** @type {?} */
    NgxSelectOption.prototype.highlightedText;
    /** @type {?} */
    NgxSelectOption.prototype.active;
    /** @type {?} */
    NgxSelectOption.prototype.cacheHighlightText;
    /** @type {?} */
    NgxSelectOption.prototype.cacheRenderedText;
    /** @type {?} */
    NgxSelectOption.prototype.value;
    /** @type {?} */
    NgxSelectOption.prototype.text;
    /** @type {?} */
    NgxSelectOption.prototype.disabled;
    /** @type {?} */
    NgxSelectOption.prototype.data;
    /** @type {?} */
    NgxSelectOption.prototype._parent;
}
var NgxSelectOptGroup = /** @class */ (function () {
    function NgxSelectOptGroup(label, options) {
        if (options === void 0) { options = []; }
        this.label = label;
        this.options = options;
        this.type = 'optgroup';
        this.filter(function () { return true; });
    }
    /**
     * @param {?} callbackFn
     * @return {?}
     */
    NgxSelectOptGroup.prototype.filter = /**
     * @param {?} callbackFn
     * @return {?}
     */
    function (callbackFn) {
        this.optionsFiltered = this.options.filter(function (option) { return callbackFn(option); });
    };
    return NgxSelectOptGroup;
}());
export { NgxSelectOptGroup };
if (false) {
    /** @type {?} */
    NgxSelectOptGroup.prototype.type;
    /** @type {?} */
    NgxSelectOptGroup.prototype.optionsFiltered;
    /** @type {?} */
    NgxSelectOptGroup.prototype.label;
    /** @type {?} */
    NgxSelectOptGroup.prototype.options;
}
/** @typedef {?} */
var TSelectOption;
export { TSelectOption };
