// ngx-select-ex/esm2015/ngx-select/ngx-select.classes.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import * as escapeStringNs from 'escape-string-regexp';
/** @type {?} */
const escapeString = escapeStringNs;
export class NgxSelectOption {
    /**
     * @param {?} value
     * @param {?} text
     * @param {?} disabled
     * @param {?} data
     * @param {?=} _parent
     */
    constructor(value, text, disabled, data, _parent = null) {
        this.value = value;
        this.text = text;
        this.disabled = disabled;
        this.data = data;
        this._parent = _parent;
        this.type = 'option';
        this.cacheRenderedText = null;
    }
    /**
     * @return {?}
     */
    get parent() {
        return this._parent;
    }
    /**
     * @param {?} sanitizer
     * @param {?} highlightText
     * @return {?}
     */
    renderText(sanitizer, highlightText) {
        if (this.cacheHighlightText !== highlightText || this.cacheRenderedText === null) {
            this.cacheHighlightText = highlightText;
            if (this.cacheHighlightText) {
                this.cacheRenderedText = sanitizer.bypassSecurityTrustHtml((this.text + '').replace(new RegExp(escapeString(this.cacheHighlightText), 'gi'), '<strong>$&</strong>'));
            }
            else {
                this.cacheRenderedText = sanitizer.bypassSecurityTrustHtml(this.text);
            }
        }
        return this.cacheRenderedText;
    }
}
if (false) {
    /** @type {?} */
    NgxSelectOption.prototype.type;
    /** @type {?} */
    NgxSelectOption.prototype.highlightedText;
    /** @type {?} */
    NgxSelectOption.prototype.active;
    /** @type {?} */
    NgxSelectOption.prototype.cacheHighlightText;
    /** @type {?} */
    NgxSelectOption.prototype.cacheRenderedText;
    /** @type {?} */
    NgxSelectOption.prototype.value;
    /** @type {?} */
    NgxSelectOption.prototype.text;
    /** @type {?} */
    NgxSelectOption.prototype.disabled;
    /** @type {?} */
    NgxSelectOption.prototype.data;
    /** @type {?} */
    NgxSelectOption.prototype._parent;
}
export class NgxSelectOptGroup {
    /**
     * @param {?} label
     * @param {?=} options
     */
    constructor(label, options = []) {
        this.label = label;
        this.options = options;
        this.type = 'optgroup';
        this.filter(() => true);
    }
    /**
     * @param {?} callbackFn
     * @return {?}
     */
    filter(callbackFn) {
        this.optionsFiltered = this.options.filter((option) => callbackFn(option));
    }
}
if (false) {
    /** @type {?} */
    NgxSelectOptGroup.prototype.type;
    /** @type {?} */
    NgxSelectOptGroup.prototype.optionsFiltered;
    /** @type {?} */
    NgxSelectOptGroup.prototype.label;
    /** @type {?} */
    NgxSelectOptGroup.prototype.options;
}
/** @typedef {?} */
var TSelectOption;
export { TSelectOption };
